### PR TITLE
A few operators on graphs stored as CSR

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
We can store a graph as an CSR matrix. In this case, the value in the matrix is the edge id starting from 0. These operators run on such special CSR matrices. This is joint work with @HyperZealot and @aksnzhy 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by zheng-da** [comment]
@eric-haibin-lin could you please review this PR?

**Comment by kalyc** [comment]
@mxnet-label-bot add [pr-awaiting-review]

File: src/operator/contrib/dgl_graph.cc [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+
+static void GetSubgraph(const NDArray &csr_arr, const NDArray &varr,
+                        const NDArray &sub_csr, const NDArray *old_eids) {
+  const TBlob &data = varr.data();
+  int64_t num_vertices = csr_arr.shape()[0];
+  const size_t len = varr.shape()[0];
+  const dgl_id_t *vid_data = data.dptr<dgl_id_t>();
+  HashTableChecker def_check(vid_data, len);
+  // check if varr is sorted.
+  std::is_sorted(vid_data, vid_data + len);
```
Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph.cc]: It returns True if it is actually sorted. But the return value is not used anywhere. Is this intended?

File: src/operator/contrib/dgl_graph.cc [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+Example::
+  x = [[ 1, 0, 0 ],
+       [ 0, 2, 0 ],
+       [ 0, 0, 3 ]]
+  u = [ 0, 0, 1, 1, 2, 2 ]
+  v = [ 0, 1, 1, 2, 0, 2 ]
+  edge_id(x, u, v) = [ 1, -1, 2, -1, -1, 3 ]
+
+The storage type of ``edge_id`` output depends on storage types of inputs
+  - quadratic(csr, default, default) = default
```
Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph.cc]: quadratic -> edge_id

File: src/operator/contrib/dgl_graph.cc [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+.set_num_outputs(1)
+.set_attr<nnvm::FListInputNames>("FListInputNames",
+  [](const NodeAttrs& attrs) {
+    return std::vector<std::string>{"data", "u", "v"};
+  })
+.set_attr<nnvm::FInferShape>("FInferShape", EdgeIDShape)
+.set_attr<nnvm::FInferType>("FInferType", EdgeIDType)
+.set_attr<FInferStorageType>("FInferStorageType", EdgeIDStorageType)
+.set_attr<FComputeEx>("FComputeEx<cpu>", EdgeIDForwardEx<cpu>)
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
```
Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph.cc]: This is actually not used since the first input is sparse. Inplace option with sparse array is not enabled

File: src/operator/contrib/dgl_graph.cc [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+  } else {
+    LogUnimplementedOp(attrs, ctx, inputs, req, outputs);
+  }
+}
+
+NNVM_REGISTER_OP(_contrib_edge_id)
+.describe(R"code(This operator implements the edge_id function for csr arrays,
+where output[i] = input[u[i], v[i]] if input[u[i], v[i]] is a non-zero element of input,
+otherwise output[i] will be -1. Both u and v should be 1D vectors.
+Example::
```
Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph.cc]: If csr.data contains zero values, would it be returned as 0 or -1? I don't see verification of "non-zero" when searching the value in csr.data

Comment (zheng-da) [reply src/operator/contrib/dgl_graph.cc]: The CSR is a special CSR. Its value is edge Id, so it contains edge id 0. So when we search for edges, we only need to check if the location has a value.

Comment (eric-haibin-lin) [reply src/operator/contrib/dgl_graph.cc]: Yes I am aware of that. That's why I think it's not appropriate to put `if input[u[i], v[i]] is a non-zero element of input` since it also returns elements that are zeros and stored in csr.
